kafka/client_async.py

```diff
@@ -612,7 +612,7 @@ def _poll(self, timeout):
             conn = key.data
             processed.add(conn)
 
-            if not conn.in_flight_requests:
+            if not conn.has_in_flight_requests():
```

Comment:
Race condition which causes the "Protocol out of sync" (the root cause) happens on this condition.